suppressPackageStartupMessages({
  library(argparse)
  library(dplyr)
  library(data.table)
  library(karyoploteR)
})

options(scipen=999)

# Calculate bp distance between two consecutive DNA motif
dist_f <- function(data) {
    data  %>%
    group_by(V1) %>%
    mutate(distance = abs(V2 - lag(V3))) %>% # Create new column for bp-distance values
    ungroup()
}

# Focus on the strand of CENP-B boxes
get_col <- function(V4) {
  if (V4 == "+") {
    return("0,0,204") # Blue to forward orientation
  } else if (V4 == "-") {
    return("204,0,0") # Red to complement reverse orientation
  } else {
    return("0,0,0") # Default to black if orientation is neither + nor -
  }
}

# Convert RGB to HEX for karyoploteR
rgb_to_hex <- function(rgb_string) {
  # Split the RGB string into numeric values
  rgb_values <- as.numeric(unlist(strsplit(rgb_string, ",")))
  # Convert to HEX using grDevices::rgb()
  grDevices::rgb(rgb_values[1], rgb_values[2], rgb_values[3], maxColorValue = 255)
}

parser <- ArgumentParser(description = "Centeny generates a BED file highlighting the strand of DNA motif occurrences")
parser$add_argument("-i", "--input", help = "DNA motif annotation file, generated by fuzznuc module", type="character", required = T)
parser$add_argument("--length", help = "Input file with the length of each chr/contigs from fuzznuc module", type="character", required = T)
parser$add_argument("--n_box", help = "Filter out sequences with number of boxes lower than INT [default 50]", type = "integer", default = 50)
parser$add_argument("-o", "--output_file", help = "Name of the output file [default centeny.bed]", type = "character", default = "centeny.bed")

# Parse args
args <- parser$parse_args()

if (is.null(args$input) || is.null(args$length)) {
  cat("Required file with the DNA motif annotation\n")
  quit(status = 1)
} else {
  cat("Reading input...\n")
  data_f <- fread(args$input)
  nbox <- (args$n_box)
  length_chr <- fread(args$length)
  output_file <- (args$output_file)
}

data_f <- data_f[order(V1, V2)]

data_fd <- dist_f(data_f) # bp-distance
cat("Calculating bp-distances...\n")
#data_fd <- na.omit(data_fd) #
data_fd$col_orient <- sapply(data_fd$V4, get_col) # color
cat("Highlighting orientation...\n")

data_fd <- data_fd[, c(1,2,3,4,5,1,2,3,6)]

#output_file <- file.path(dirname(args$input), "centeny.bed")
#write.table(data_fd, output_file, sep="\t", col.names = F, row.names = F, quote = F)
write.table(data_fd, output_file, sep = "\t", col.names = F, row.names = F, quote = F)

# Preparing the data for plot
cat(sprintf("Filtering out chr/contigs with n. of boxes less than %d\n", nbox))
#n_box <- 500
data_fd <- data_fd[, c(1,2,3,4,5,9)]
data_fd <- data_fd %>%
  group_by(V1) %>%
  filter(n() > nbox) %>% #minimum n of boxes within the read
  arrange(V1, V2) %>%
  as.data.table()

data_fd$identificator <- "Mb"
customGenome <- toGRanges(data.frame(chr= "Mb" , start= 0, end = max(length_chr$V2)))
data_final <- data_fd[, c(7,2,3,4,5,6,1)]

# Apply the conversion to the V9 column
data_final$HEX <- sapply(data_final$col_orient, rgb_to_hex)
data_final <- split(data_final, data_final$V1)

# Data to plot the length of each chr/contigs
length_chr$V3 <- "0"
length_chr$V4 <- "Mb"
length_chr <- length_chr[, c(4,3,2,1)]

length_chr <- length_chr %>%
  filter(V1 %in% data_fd$V1) %>% #filtering contig's length by contigs with n_box > n_box
  arrange(V1) %>%
  as.data.table()
length_chr <- split(length_chr, length_chr$V1)

output_file_plot <- file.path(dirname(args$output_file), "centeny.jpeg")
#pdf(output_file_plot, width = 8000, height = 8000)
#tiff(output_file_plot, res = 1200, width = 8000, height = 8000)
cat("Preparing the centeny plot...\n")
#tiff(output_file_plot, res = 1200, width = 8000, height = 8000)
jpeg(output_file_plot, res = 300, width = 8000, height = 8000, quality = 100)

pp <- getDefaultPlotParams(plot.type=1)
pp$ideogramheight <- 0
pp$ideogramlateralmargin <- 0.008
pp$data2height <- 0
pp$data1inmargin <- 0
pp$data2inmargin <- 0
pp$data1outmargin <- 0
pp$data1height <- 5000 #1200
pp$leftmargin <- 0.15 #0.4 for the panel
pp$rightmargin <- 0.15
pp$topmargin <- 800
pp$data2height <- 50
pp$data1max <- 1.5 #2
pp$bottommargin <- 150

kp <- plotKaryotype(plot.type=1, genome = customGenome, plot.params=pp, cex=0.4, labels.plotter = NULL, chromosomes="all")
kp <- kpAddBaseNumbers(kp, tick.dist = 10000000, tick.len = 10, tick.col="black", cex=1, minor.tick.dist = 5000000, minor.tick.col = "black", minor.tick.len = 4, units = "Mb")

initial_r0 <- 0.0
initial_r1 <- 0.013
increment <- 0.02

process <- lapply(seq_along(data_final), function(i) {
  current_r0 <- initial_r0 + increment * i
  current_r1 <- initial_r1 + increment * i

  current_gr <- toGRanges(as.data.frame(data_final[[i]]))

  kp <- kpPlotRegions(kp, data.panel = 1, r0 = current_r1, r1 = current_r0, data = current_gr, col = current_gr$HEX)
  kp <- kpAddLabels(kp, labels = current_gr$V1, r0 = current_r1, r1 = current_r0, data.panel = 1, cex = 1, side = "left", label.margin = 0.01, font = 1)
})

initial_r0 <- 0.0
initial_r1 <- 0.000001
increment <- 0.02

process <- lapply(seq_along(length_chr), function(i) {
  current_r0 <- initial_r0 + increment * i
  current_r1 <- initial_r1 + increment * i

  current_gr <- toGRanges(as.data.frame(length_chr[[i]]))

  kp <- kpPlotRegions(kp, data.panel = 1, r0 = current_r1, r1 = current_r0, data = current_gr, col = "black")
  #kpAddLabels(kp, labels = current_gr$V1, r0 = current_r1, r1 = current_r0, data.panel = 1, cex = 0.35, side = "left", label.margin = 0.01, font = 1)
})

invisible(dev.off())