#!/usr/bin/env Rscript
library(argparse)

# Define arguments
parser <- ArgumentParser(description = "This tool handles the annotation of DNA motif (CENP-B box by default) occurrences along the genome to characterize the chromosome-level organization")
subparsers <- parser$add_subparsers(dest = "command", help = "Available commands")

# Search for DNA motif within FASTA file
parser_fuzznuc <- subparsers$add_parser("fuzznuc", help = "Fuzznuc searches for a specific DNA motif within a FASTA file. This step produces the input for all the other commands")
parser_fuzznuc$add_argument("-fa" , "--fasta_file", required = TRUE, help = "FASTA file where to search for the DNA motif", type = "character", metavar = "FILE")
parser_fuzznuc$add_argument("-seq" , "--sequence", help = "DNA motif to search within the FASTA. By default it searches for the CENP-B box [default NTTCGNNNNANNCGGGN]", type = "character", default = "NTTCGNNNNANNCGGGN")
parser_fuzznuc$add_argument("-o", "--output_dir", help = "Path of the output directory [default .]", type = "character", metavar = "DIR_NAME")

# Define the 3 models
# Centeny
parser_centeny <- subparsers$add_parser("centeny", help = "Centeny generates a BED file highlighting the strand of DNA motif occurrences")
parser_centeny$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type = "character", metavar = "FILE")
parser_centeny$add_argument("--length", required = TRUE, help = "Input file with the length of each chr/contigs from fuzznuc module", type="character", metavar = "FILE")
parser_centeny$add_argument("--n_box", help = "Filter out sequences with number of boxes lower than n_box [default 50]", metavar = "INT", type = "integer", default = 50)
parser_centeny$add_argument("-o", "--output_file", help = "Name of the output file [default centeny.bed]", type = "character", default = "centeny.bed", metavar = "FILE")

# MODEL1
parser_model1 <- subparsers$add_parser("model1", help = "Model1 generates a BED file highlighting the conservation of bp-distances with respect to CHM13")
parser_model1$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type = "character", metavar = "FILE")
parser_model1$add_argument("-d", "--distance", required = TRUE, help = "Tab-separated file with distance values represented more than 1 percent in at least one chromosome of the CHM13 (distance column) and associated colors (color column)", type = "character", metavar = "FILE")
parser_model1$add_argument("-o", "--output_file", help = "Name of the output file [default model1.bed]", type = "character", default = "model1.bed", metavar = "FILE")

# MODEL2
parser_model2 <- subparsers$add_parser("model2", help = "Model2 generates a BED file highlighting the conservation of the motif organization with respect to alpha-satellite monomers")
parser_model2$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type = "character", metavar = "FILE")
parser_model2$add_argument("-d", "--distance", required = TRUE, help = "Tab-separated file where the distance column contains the number of monomers between two monomers with the DNA motif and the color column reports the associated color", type = "character", metavar = "FILE")
parser_model2$add_argument("-o", "--output_file", help = "Name of the output file [default model2.bed]", type = "character", default = "model2.bed", metavar = "FILE")

# MODEL3
parser_model3 <- subparsers$add_parser("model3", help = "Model3 handles the annotation of motif occurrences along the genome by annotating their organization using k-pattern")
parser_model3$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type="character", metavar = "FILE")
parser_model3$add_argument("-k", "--klength", help = "Number of distance values includeed in each pattern [default 4]", type="integer", default = 4, metavar = "INT")
parser_model3$add_argument("-fq", "--frequency", help = "Minimum frequency required to color a k-pattern [default 5]", type="integer", default = 5, metavar = "INT")
parser_model3$add_argument("-o", "--output_file", help = "Name of the output file [default model3.bed]", type = "character", default = "model3.bed", metavar = "FILE")

# Reads Retrieval
parser_reads_retrieval <- subparsers$add_parser("reads_retrieval", help = "Reads_retrieval isolates chromosome-specific centromeric reads from a FASTA/FASTQ file by using chromosome-specific kpatterns from model3")
parser_reads_retrieval$add_argument("-i", "--input", help = "DNA motif annotation file, generated by applying the fuzznuc module to long sequencing reads", type = "character", required = TRUE, metavar = "FILE")
parser_reads_retrieval$add_argument("-p", "--pattern_table", help = "Input table from model3 which stores all the k-patterns with a sliding step of 1", type = "character", required = TRUE, metavar = "FILE")
parser_reads_retrieval$add_argument("-c", "--chr", help = "Chromosome of interest (names are stored in column 3 of the file in pattern_table)", type = "character", required = TRUE, metavar = "CHR_NAME")
parser_reads_retrieval$add_argument("--n_box", help = "Filter out reads with number of motif occurrences lower than INT (this parameter must to be equal to/higher than the number of distances in the k-pattern)", type = "integer", required = TRUE, metavar = "INT")
parser_reads_retrieval$add_argument("--n_pattern", help = "Consider only k-patterns with frequency higher than INT within the selected chromosome [default 10]", type = "integer", default = 10, metavar = "INT")
parser_reads_retrieval$add_argument("--n_matches", help = "Assign the read to the selected chromosome if it contains at least INT considered k-patterns [default 1]", type = "integer", default = 1, metavar = "INT")
parser_reads_retrieval$add_argument("-t", "--n_cores", help = "Number of threads [default 5]", type = "integer", default = 5, metavar = "INT")
parser_reads_retrieval$add_argument("-o", "--output_file", help = "Name of the output file [default chr_specific_reads.txt]", type = "character", default = "chr_specific_reads.txt", metavar = "FILE")

# Parse arguments
args <- parser$parse_args()

if (is.null(args$command)) {
  parser$print_help()
  stop("Error: No command specified.")
}

# Dispatch to the appropriate script
if (args$command == "fuzznuc") {
  if (file.exists("fuzznuc.sh")) {
    cmd <- sprintf("bash fuzznuc.sh %s %s %s", args$fasta_file, args$sequence, args$output_dir)
    system(cmd)
  } else {
    stop("Error: 'fuzznuc.sh' not found.")
  }
} else if (args$command == "centeny") {
  if (file.exists("centeny.R")) {
    cmd <- sprintf("Rscript centeny.R -i %s --length %s --n_box %d -o %s", args$input, args$length, args$n_box, args$output_file)
    system(cmd)
  } else {
    stop("Error: 'centeny.R' not found.")
  }
} else if (args$command == "model1") {
  if (file.exists("model1.R")) {
    cmd <- sprintf("Rscript model1.R -i %s -d %s -o %s", args$input, args$distance, args$output_file)
    system(cmd)
  } else {
    stop("Error: 'model1.R' not found.")
  }
} else if (args$command == "model2") {
  if (file.exists("model2.R")) {
    cmd <- sprintf("Rscript model2.R -i %s -d %s -o %s", args$input, args$distance, args$output_file)
    system(cmd)
  } else {
    stop("Error: 'model2.R' not found.")
   }
} else if (args$command == "model3") {
  if (file.exists("model3.R")) {
    cmd <- sprintf("Rscript model3.R -i %s -k %d -fq %d -o %s", args$input, args$klength, args$frequency, args$output_file)
    system(cmd)
  } else {
    stop("Error: 'model3.R' not found.")
   }
} else if (args$command == "reads_retrieval") {
  if (file.exists("reads_retrieval.R")) {
    cmd <- sprintf("Rscript reads_retrieval.R -i %s -p %s --chr %s --n_box %d --n_pattern %d --n_matches %d --n_cores %d -o %s", args$input, args$pattern_table, args$chr, args$n_box, args$n_pattern, args$n_matches, args$n_cores, args$output_file)
    system(cmd)
  } else {
    stop("Error: 'reads_retrieval.R' not found.")
   }
} else {
  parser$print_help()
  stop("Error: Unknown command.")
}