suppressPackageStartupMessages({
  library(argparse)
  library(dplyr)
  library(data.table)
})

options(scipen=999)

# Calculate bp distance between two consecutive DNA motif
dist_f <- function(data) {
    data  %>%
    group_by(V1) %>%
    mutate(distance = abs(V2 - lag(V3))) %>% # Create new column for bp-distance values
    ungroup()
}

# Define a function to get color based on distance
get_color_from_distance <- function(distance) {
  if (is.na(distance)) {
    return("0,0,0")  # Return black if the distance is NA, keeping the NA for IGV visualization of the first CENP-B box of the chromosome
  }
  color_row <- distance_f[distance_f$distance == distance, "color"]
  if (length(color_row) > 0) {
    return(as.character(color_row))
  } else {
    return("0,0,0")  # Return black for distances not found in the color file / <1% represented
  }
}

parser <- ArgumentParser(description = "Model1 generates a BED file highlighting the conservation of bp-distances with respect to CHM13")
parser$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type = "character", metavar = "FILE")
parser$add_argument("-d", "--distance", required = TRUE, help = "Tab-separated file with distance values represented more than 1 percent in at least one chromosome of the CHM13 (distance column) and associated colors (color column)", type = "character", metavar = "FILE")
parser$add_argument("-o", "--output_file", help = "Name of the output file [default model1.bed]", type = "character", default = "model1.bed", metavar = "FILE")

# Parse args
args <- parser$parse_args()

# Check if both parameters are provided
if (is.null(args$input) || is.null(args$distance)) {
  cat("HELP: Both '--input' (DNA motif annotation file), '--distance' (distance and color file) and '--length (length of each chr/contigs) are required!\n")
  quit(status = 1) # Exit the script with an error status
} else {
  cat("Reading input files...\n")
  data_f <- fread(args$input)
  distance_f <- fread(args$distance)
  output_file <- (args$output_file)
}

distance_f <- as.data.frame(distance_f)
data_f <- data_f[order(V1, V2)]

# Creating list of dataframe by chromosome
data_f <- split(data_f, data_f$V1)

cat("Calculating bp-distances...\n")
data_fd <- lapply(data_f, dist_f) # bp-distance
#data_fd <- na.omit(data_fd) #to remove all the NA distance values (NA values belong to the first CENP-B box of the chromosome)


cat("Assigning colors to distances...\n")
# Assign color to distance value
for (i in seq_along(data_fd)) {
  # Add a new column 'color' to each table
  data_fd[[i]]$color <- sapply(data_fd[[i]]$distance, get_color_from_distance)
}

data_fd <- bind_rows(data_fd)

# Giving a name to the distance values
data_fd$distance <- paste0("D", data_fd$distance)

data_fd <- data_fd[, c(1,2,3,5,4,1,2,3,6)]

#output_file <- file.path(dirname(args$input), "model1.bed")
write.table(data_fd, output_file, sep="\t", col.names = F, row.names = F, quote = F)