suppressPackageStartupMessages({
  library(argparse)
  library(dplyr)
  library(data.table)
  library(randomcoloR)
})

options(scipen=999)

# Calculate bp distance between two consecutive DNA motif
dist_f <- function(data) {
  data  %>%
    group_by(V1) %>%
    mutate(distance = abs(V2 - lag(V3))) %>% # Create new column for bp-distance values
    ungroup()
}

create_patterns_slidingk <- function(df, n) {
  chromosome <- df$V1
  start <- df$V2
  end <- df$V3
  orientation <- df$V4
  series_of_numbers <- as.numeric(df$distance)
  # Annotating k-pattern with sliding window of n
  patterns <- lapply(seq(1, length(series_of_numbers) - n + 1, by = n), function(i) {
    list(
      pattern = series_of_numbers[i:(i + n - 1)],
      chromosome = chromosome[1],
      start = start[i],
      end = end[i + n - 1],
      orientation = orientation[1]
    )
  })

  # Convert list to a data frame
  patterns_df <- do.call(rbind, lapply(patterns, function(x) {
    data.frame(
      chromosome = x$chromosome,
      start = x$start,
      end = x$end,
      orientation = x$orientation,
      pattern = paste0("#", paste(x$pattern, collapse = "_"))
    )
  }))

  return(patterns_df)
}

# Color assignment to k-pattern
assign_colors <- function(data) {
  # Count occurrences of each unique value (k-pattern) in V5
  freq_table <- table(data$pattern)

  # Filter for values that appear more than 1 times
  frequent_values <- names(freq_table[freq_table > fq])

  # Generate a palette of colors for frequent values
  col <- length(frequent_values)
  colorpal <- distinctColorPalette(col) # HEX colors

  # Convert HEX to RGB
  rgb_colors <- apply(col2rgb(colorpal), 2, function(rgb) paste(rgb, collapse = ",")) # Convert to "R,G,B" format

  # Create a map of RGB colors for frequent values
  samp_col <- setNames(rgb_colors, frequent_values)

  # Assign RGB colors to frequent values, and default color to others (e.g., "0,0,0" for black)
  data$col <- ifelse(data$pattern %in% frequent_values, samp_col[data$pattern], "0,0,0")

  # Convert the 'col' column to a factor (optional)
  data$col <- as.factor(data$col)

  return(data)
}

parser <- ArgumentParser(description = "Model3 handles the annotation of motif occurrences along the genome by annotating their organization using k-pattern")
parser$add_argument("-i", "--input", required = TRUE, help = "DNA motif annotation file, generated by fuzznuc module", type="character", metavar = "FILE")
parser$add_argument("-k", "--klength", help = "Number of distance values includeed in each pattern [default 4]", type="integer", default = 4, metavar = "INT")
parser$add_argument("-fq", "--frequency", help = "Minimum frequency required to color a k-pattern [default 5]", type="integer", default = 5, metavar = "INT")
parser$add_argument("-o", "--output_file", help = "Name of the output file [default model3.bed]", type = "character", default = "model3.bed", metavar = "FILE")
args <- parser$parse_args()


if (is.null(args$input)) {
  cat("Required input with the DNA motif annotation\n")
  quit(status = 1) # Exit the script with an error status
} else {
  cat("Reading input file...\n")
  df_f <- fread(args$input)
  n <- (args$klength)
  fq <- (args$frequency)
  output_file_final <- (args$output_file)
}

df_f <- df_f[order(V1, V2)]

cat("Calculating bp-distances...\n")
df_f <- dist_f(df_f)
df_f <- na.omit(df_f)

# Split based on chr/read/contig names
df_f <- split(df_f, df_f$V1)

#Avoid contigs with less box than range_k
df_f <- Filter(function(df) nrow(df) >= max(n), df_f)

# Column of bp-distance by chr/read/contig into a numeric vector
df_ff <- lapply(df_f, function(data) {
  data <- as.numeric(data$distance)
  return(data)
})

# SLIDING 1
cat("Initializing k-pattern with sliding of 1...\n")
patterns <- list() # initializes an empty list

# loops over the indices of the list
# generate complete subsequences of length n starting at index j with a step size of 1 (sliding window of 1)
# extracts a subsequence from the j-th element to the (j + n - 1)-th element of the vector df_ff[[i]]
for (i in seq_along(df_ff)) {
  patterns[[i]] <- lapply(1:(length(df_ff[[i]])-n+1), function(j) {
    df_ff[[i]][j:(j+n-1)]
  })
}

names(patterns) <- names(df_ff)

# Counting the frequency of each k-pattern
pattern_counts <- lapply(patterns, function(x) {
  f <- table(as.character(x))
  d <- t(t(f))
  data.frame(d[order(d[,1], decreasing = T), ])
})

pattern_counts <- lapply(names(pattern_counts), function(seq_name) {
  lapply(seq_along(pattern_counts[[seq_name]]), function(n) {
    # Extract the pattern count table
    df <- pattern_counts[[seq_name]]
    df <- as.data.frame(df)  # Ensure it's a data frame if it's not already
    # Add new columns: row names, frequency (V2), sequence name (V3)
    df$V1 <- rownames(df)  # Row names as the first column
    df$V2 <- df[, 1]       # Frequency as the second column
    df$V3 <- seq_name      # Add sequence name as the third column
    df <- df[, c("V1", "V2", "V3")]  # Keep only V1 (pattern), V2 (frequency), V3 (sequence name)
    rownames(df) <- NULL # Remove row names
    return(df)  # Return the cleaned data frame
  })
})

pattern_counts <- bind_rows(pattern_counts)

# Save k-pat in one single file
output_dir <- dirname(output_file_final)
output_file <- file.path(output_dir, paste0("pattern_counts_sliding1_k", n, '.txt'))
write.table(pattern_counts, output_file, quote = F, sep = "\t", col.names = F, row.names = F)

cat(sprintf("Table of patterns with sliding 1 of length %d\n", n))

# SLIDING k ANNOTATION
cat(sprintf("Annotation of consecutive patterns with sliding k of length %d\n", n))
# Running the function to annotate k-pattern with sliding window of k
patterns_list <- lapply(df_f, create_patterns_slidingk, n)
patterns_df_k <- bind_rows(patterns_list) #from list to dataframe
# Running the function to assign colors to unique k-pattern with frequency > 1
cat(sprintf("Assigning random colors to k-patterns with frequency higher than %d\n", fq))
patterns_df_k <- assign_colors(patterns_df_k)
patterns_df_k$score <- '.'
# Formatting the table
patterns_df_k <- patterns_df_k[,c(1,2,3,5,7,4,2,3,6)]
#output_file <- file.path(dirname(args$input), paste0("model3_slidingk_annotation_k", n, '.bed')) #paste0("final_annotation_slidingk_annotation_k", n, '.bed')
write.table(patterns_df_k, output_file_final, sep="\t", col.names = F, row.names = F, quote = F)